src/GSMConnectionManager.h

```diff
@@ -26,50 +26,175 @@ class GSMConnectionManager : public ConnectionManager {
   virtual void init();
   virtual void check();
   virtual Client &getClient() { return networkClient; };
+  virtual UDP &getUDP() { return udp; };
 
   GSMClient networkClient;
   GSM gsmAccess;
   GPRS gprs;
+  GSMUDP udp;
 private:
+
+  void changeConnectionState(NetworkConnectionState _newState);
+
+  const int CHECK_INTERVAL_IDLE = 100;
+  const int CHECK_INTERVAL_INIT = 100;
+  const int CHECK_INTERVAL_CONNECTING = 500;
+  const int CHECK_INTERVAL_GETTIME = 1000;
+  const int CHECK_INTERVAL_CONNECTED = 10000;
+  const int CHECK_INTERVAL_RETRYING = 5000;
+  const int CHECK_INTERVAL_DISCONNECTED = 1000;
+  const int CHECK_INTERVAL_ERROR = 500;
+
   const char *pin, *apn, *login, *pass;
-  unsigned long lastNetworkCheck, lastNetworkStep;
+  unsigned long lastConnectionTickTime, lastNetworkStep;
+  int connectionTickTimeInterval;
+  GSMUDP Udp;
+  void sendNTPpacket(const char * address, uint8_t* packetBuffer);
+  unsigned long getNTPTime();
 };
 
 static const unsigned long NETWORK_CONNECTION_INTERVAL = 30000;
 
 GSMConnectionManager::GSMConnectionManager(const char *pin, const char *apn, const char *login, const char *pass) :
```

Comment:
The member variables `pin`, `apn`, `login` and `pass` are only stored as const char pointers. This implies that those variables **must be** of global scope within the calling code, otherwise we might read invalid data if we read them at a later point in time. A alternative would be to copy the content of those variables into the class, e.g. via a `String` data type.

Reply:
I agree, the structure is the same as WiFi wrapper that "requires" variables with global scope. I'd be ok to switch to strings, but it should be something shared between all the wrappers.

Reply:
@lxrobotics @facchinm 
because we use them in the Web IDE to create fields in the `Secret` tab, they are indeed globally defined for our cases.
storing them as String might make sense. I also side with @facchinm with having this kind of structure shared amongst the wrappers